pkg/services/email.go

```diff
@@ -0,0 +1,175 @@
+package services
+
+import (
+	"errors"
+	"github.com/kptm-tools/core-service/pkg/customerrors"
+	"github.com/kptm-tools/core-service/pkg/domain"
+	"github.com/kptm-tools/core-service/pkg/interfaces"
+	gomail "gopkg.in/mail.v2"
+	"log/slog"
+	"os"
+	"strconv"
+	"time"
+)
+
+const (
+	maxRetries        = 3
+	initialRetryDelay = 5 * time.Second
+)
+
+type SendMailFunction func(...*gomail.Message) error
+
+type EmailService struct {
+	Host      string
+	Port      string
+	Username  string
+	Password  string
+	FromEmail string
+	SendMail  SendMailFunction
+}
+
+var _ interfaces.IEmailService = (*EmailService)(nil)
+
+func NewEmailService(host, port, username, password, fromEmail string) *EmailService {
+	portNum, _ := strconv.ParseInt(port, 0, 0)
+	dialer := gomail.Dialer{
+		Host:           host,
+		Port:           int(portNum),
+		Username:       username,
+		Password:       password,
+		StartTLSPolicy: gomail.NoStartTLS,
+	}
+
+	return &EmailService{
+		Host:      host,
+		Port:      port,
+		Username:  username,
+		Password:  password,
+		FromEmail: fromEmail,
+		SendMail:  dialer.DialAndSend,
+	}
+}
+
+func (s *EmailService) SendEmail(toAddress []*domain.Rapporteur, subject, body string) error {
+	if toAddress == nil || len(toAddress) == 0 {
+		return errors.New("no emails configured to be sent")
+	}
+	m := gomail.NewMessage()
+	sizeAddress := len(toAddress)
+	addresses := make([]string, sizeAddress)
+	for i, recipient := range toAddress {
+		addresses[i] = m.FormatAddress(recipient.Email, recipient.Name)
+	}
+	m.SetHeader("From", s.FromEmail)
+	m.SetHeader("To", addresses...)
+	m.SetHeader("Subject", subject)
+	m.SetBody("text/html", body)
+
+	if err := s.SendMail(m); err != nil {
+		return err
+	} else {
+		slog.Info("Email sent")
+	}
+	return nil
+}
+
+func (s *EmailService) SendScanFailedEmail(recipient string, hostName string) error {
+	subject := "Scan Failed of" + hostName
+	body := "Dear Recipient, the scan has been failed"
+	recipientRapporteur := []*domain.Rapporteur{
+		{
+			"",
+			recipient,
+			false,
+		},
+	}
+	err := s.sendEmailWithRetry(recipientRapporteur, subject, body)
+	if err != nil {
+		return err
+	}
+	slog.Info("Email of Scan failed sent", slog.String("recipient", recipient))
+	return nil
+}
+
+func (s *EmailService) SendScanCompletedEmail(recipient string, hostName string) error {
+	subject := "Scan Completed of " + hostName
+	body := "Dear Recipient, the scan has been completed"
+	recipientRapporteur := []*domain.Rapporteur{
```

Comment:
These struct literal is using unkeyed fields. While this is not a compile time error, it does have bad code smell and reduces readability, please use keys when defining struct literals:
```suggestion
	recipientRapporteur := []*domain.Rapporteur{
	  Name: "",
	  Email: recipient,
	  IsPrincipal: false,
	 }
```